tests/testsuite/build_dir.rs

```diff
@@ -0,0 +1,589 @@
+//! Tests for `build.build-dir` config property.
+//!
+//! The testing strategy for build-dir functionality is primarily checking if directories / files
+//! are in the expected locations.
+//! The rational is that other tests will verify each individual feature, while the tests in this
+//! file verify the files saved to disk are in the correct locations according to the `build-dir`
+//! configuration.
+//!
+//! Tests check if directories match some "layout" by using [`assert_build_dir_layout`] and
+//! [`assert_artifact_dir_layout`].
+
+use std::path::PathBuf;
+
+use cargo_test_support::prelude::*;
+use cargo_test_support::project;
+use std::env::consts::{DLL_PREFIX, DLL_SUFFIX, EXE_SUFFIX};
+
+#[cargo_test]
+fn verify_build_dir_is_disabled_by_feature_flag() {
+    let p = project()
+        .file("src/main.rs", r#"fn main() { println!("Hello, World!") }"#)
+        .file(
+            ".cargo/config.toml",
+            r#"
+            [build]
+            build-dir = "build"
+            "#,
+        )
+        .build();
+
+    p.cargo("build")
+        .masquerade_as_nightly_cargo(&[])
+        .enable_mac_dsym()
+        .run();
+
+    assert_build_dir_layout(p.root().join("target"), "debug");
+    assert_exists(&p.root().join(format!("target/debug/foo{EXE_SUFFIX}")));
+    assert_exists(&p.root().join("target/debug/foo.d"));
+    assert_not_exists(&p.root().join("build"));
+}
+
+#[cargo_test]
+fn binary_with_debug() {
+    let p = project()
+        .file("src/main.rs", r#"fn main() { println!("Hello, World!") }"#)
+        .file(
+            ".cargo/config.toml",
+            r#"
+            [build]
+            target-dir = "build"
+            "#,
+        )
+        .build();
+
+    p.cargo("build")
+        .masquerade_as_nightly_cargo(&[])
+        .enable_mac_dsym()
+        .run();
+
+    assert_build_dir_layout(p.root().join("build"), "debug");
+    assert_exists_patterns_with_base_dir(
+        &p.root(),
+        &[
+            &format!("build/debug/deps/foo*{EXE_SUFFIX}"),
+            "build/debug/deps/foo*.d",
+            "build/debug/.fingerprint/foo-*/bin-foo",
+            "build/debug/.fingerprint/foo-*/bin-foo.json",
+            "build/debug/.fingerprint/foo-*/dep-bin-foo",
+            "build/debug/.fingerprint/foo-*/invoked.timestamp",
```

Comment:
We probably only need to check where `.fingerprint/` is written to, not every class of file inside of it